libcrux-ml-kem/src/polynomial.rs

```diff
@@ -0,0 +1,366 @@
+//use crate::hax_utils::hax_debug_assert;
+use super::arithmetic::*;
+use super::intvec::*;
+
+pub(crate) const VECS_IN_RING_ELEMENT: usize =
+    super::constants::COEFFICIENTS_IN_RING_ELEMENT / SIZE_VEC;
+
+#[derive(Clone, Copy)]
+pub struct PolynomialRingElement {
+    pub(crate) coefficients: [IntVec; VECS_IN_RING_ELEMENT],
+}
+
+impl PolynomialRingElement {
+    pub(crate) const ZERO: Self = Self {
+        coefficients: [ZERO_VEC; VECS_IN_RING_ELEMENT],
+    };
+}
+
+#[inline(always)]
+pub(crate) fn to_i32_array(a: PolynomialRingElement) -> [i32; 256] {
+    let mut result = [0i32; 256];
+    for i in 0..VECS_IN_RING_ELEMENT {
+        let result_i = int_vec_to_i32_array(a.coefficients[i]);
+        for j in 0..SIZE_VEC {
+            result[i * SIZE_VEC + j] = result_i[j];
+        }
+    }
+    result
+}
+
+#[inline(always)]
+pub(crate) fn from_i32_array(a: [i32; 256]) -> PolynomialRingElement {
+    let mut result = PolynomialRingElement::ZERO;
+    for i in 0..VECS_IN_RING_ELEMENT {
+        result.coefficients[i] =
+            int_vec_from_i32_array(a[i * SIZE_VEC..(i + 1) * SIZE_VEC].try_into().unwrap());
+    }
+    result
+}
+
+/// Given two polynomial ring elements `lhs` and `rhs`, compute the pointwise
+/// sum of their constituent coefficients.
+#[inline(always)]
+pub(crate) fn add_to_ring_element<const K: usize>(
+    mut lhs: PolynomialRingElement,
+    rhs: &PolynomialRingElement,
+) -> PolynomialRingElement {
+    for i in 0..lhs.coefficients.len() {
+        lhs.coefficients[i] = add_int_vec(lhs.coefficients[i], &rhs.coefficients[i]);
+    }
+    lhs
+}
+
+#[inline(always)]
+pub(crate) fn poly_barrett_reduce(mut a: PolynomialRingElement) -> PolynomialRingElement {
+    for i in 0..VECS_IN_RING_ELEMENT {
+        a.coefficients[i] = barrett_reduce_int_vec(a.coefficients[i]);
```

Comment:
Should we be using `map` instead of these loops? Would make proofs simpler.

Reply:
When hax and eurydice are happy with map, it would be more rusty if we used `map`.